src/Components/OATPropertyEditor/Internal/EditorJsonTab/Internal/JSONEditor/JSONEditor.tsx

```diff
@@ -155,85 +133,44 @@ const JSONEditor: React.FC<IJSONEditorProps> = (props) => {
                 });
             };
 
-            if (parsingError) {
-                oatPageDispatch({
-                    type: OatPageContextActionType.SET_OAT_ERROR,
-                    payload: {
-                        title: t('OATPropertyEditor.errorInvalidJSON'),
-                        message: parsingError
-                    }
-                });
-            } else {
-                execute(save, undoSave);
-            }
+            execute(save, undoSave);
         };
 
-        const updatedItem = isJsonStringValid(content);
-        if (!updatedItem) {
-            oatPageDispatch({
-                type: OatPageContextActionType.SET_OAT_ERROR,
-                payload: {
-                    title: t('OATPropertyEditor.errorInvalidJSON'),
-                    message: t('OATPropertyEditor.errorInvalidJSONMessage')
-                }
-            });
-            return;
-        }
-        let updatedModel: DtdlInterface;
-        const originalModelId = oatPageState.selection.modelId;
-        if (isDTDLModel(updatedItem)) {
-            // bind the updated model
-            updatedModel = updatedItem;
-        } else if (isDTDLReference(selectedItem)) {
-            // get the model and update the reference on the model
-            updatedModel = getModelById(
-                oatPageState.currentOntologyModels,
-                originalModelId
-            );
-            const contents = ensureIsArray(updatedModel.contents);
-            const index = getReferenceIndexByName(
-                updatedModel,
-                selectedItem.name // use the original name in case they change it in the update
-            );
-            contents[index] = updatedItem;
-            updatedModel.contents = contents;
-        }
-
-        // validate the updated collection is valid
-        const models = oatPageState.currentOntologyModels;
-        const modelIndex = getModelIndexById(models, originalModelId);
-        models[modelIndex] = updatedModel;
-        const parsingError = await parseModels(models);
+        const validationResult = await validateItemChangeBeforeSaving({
+            content: content,
+            originalItem: selectedItem,
+            existingModels: oatPageState.currentOntologyModels,
+            selectedModelId: oatPageState.selection.modelId
+        });
 
-        if (parsingError) {
+        if (validationResult.isValid === true) {
```

Comment:
is there a reason why it is not only `if (validationResult.isValid)`

Reply:
Yeah. I don't know why but typescript only figures out the discriminated union if I do the explicit check. I feel like it should be smart enough but alas I haven't figured out an alternative here.